### PR TITLE
FIPS Mode for RHEL x86

### PR DESCRIPTION
**Pull request by WilburZjh**
Signed-off-by: Jinhang Zhang <Jinhang.Zhang@ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
The copyright check failed
```
12:28:13  make/modules/java.base/Copy.gmk
12:28:13  	IBM Copyright (basic) is missing from the file
```

**Comment by WilburZjh** [comment]
> 



> The copyright check failed
> 
> ```
> 12:28:13  make/modules/java.base/Copy.gmk
> 12:28:13  	IBM Copyright (basic) is missing from the file
> ```

Fixed

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 1]
```diff
@@ -109,223 +111,269 @@ public final class SunEntries {
...
-         * Set the digest provider based on whether native crypto is
-         * enabled or not.
-         */
-        if (useNativeDigest) {
-            providerSHA = "sun.security.provider.NativeSHA";
-            providerSHA224 = "sun.security.provider.NativeSHA2$SHA224";
-            providerSHA256 = "sun.security.provider.NativeSHA2$SHA256";
-            providerSHA384 = "sun.security.provider.NativeSHA5$SHA384";
-            providerSHA512 = "sun.security.provider.NativeSHA5$SHA512";
+        if(FIPSConfigurator.enableFips()) {
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java:114]: Space after `if`, please.

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java [hunk 2]
```diff
@@ -199,135 +207,185 @@ public Void run() {
...
-            names.append(namedCurve.getObjectId());
-            names.append("]");
-        }
-
-        HashMap<String, String> apAttrs = new HashMap<>(ATTRS);
-        apAttrs.put("SupportedCurves", names.toString());
-
-        putService(new ProviderServiceA(this, "AlgorithmParameters",
-            "EC", "sun.security.util.ECParameters", apAttrs));
+        if (FIPSConfigurator.enableFips()) {
```
Comment (keithc-ca) [inline src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:210]: Both branches of this `if` statement appear to be the same. Shouldn't they be different?

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: closed/src/java.base/share/classes/openj9/internal/security/FIPSConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        } catch (Exception e) {
+            if (debug != null) {
+                debug.println("Unable to load FIPS configuration");
+                e.printStackTrace();
+            }
+        }
+        return loadedProps;
+    }
+}
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply closed/src/java.base/share/classes/openj9/internal/security/FIPSConfigurator.java]: added

File: make/modules/java.base/Copy.gmk [hunk 4]
```diff
@@ -266,3 +269,15 @@ $(eval $(call SetupTextFileProcessing, CREATE_CLASSFILE_CONSTANTS_H, \
...
+
+ifeq ($(OPENJDK_TARGET_OS)-$(OPENJDK_TARGET_CPU_ARCH), linux-x86)
+  NSS_FIPS_CFG_SRC := $(TOPDIR)/closed/src/java.base/share/conf/security/nss.fips.cfg
+  NSS_FIPS_CFG_DST := $(CONF_DST_DIR)/security/nss.fips.cfg
+
+  $(NSS_FIPS_CFG_DST) : $(NSS_FIPS_CFG_SRC)
+	$(call install-file)
+
+  TARGETS += $(NSS_FIPS_CFG_DST)
+endif
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply make/modules/java.base/Copy.gmk]: changed to `closed/custom/modules/java.base/Copy.gmk`

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java (hunk 2 above)
Comment (WilburZjh) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:210]: same as `SunEntries.java`

File: closed/src/java.base/share/classes/openj9/internal/security/FIPSConfigurator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply closed/src/java.base/share/classes/openj9/internal/security/FIPSConfigurator.java]: Please fix all files; each of these don't end with a newline:
```
closed/custom/modules/java.base/Copy.gmk
closed/src/java.base/share/conf/security/nss.fips.cfg
test/jdk/ProblemList-fips.txt
```

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java (hunk 2 above)
Comment (keithc-ca) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:210]: Perhaps you don't understand my question. What I see has this pattern:
```
    if (FIPSConfigurator.enableFips()) {
        /* block A */
    } else {
        /* block B */
    }
```
with block A and block B doing exactly the same thing. If this file is unchanged, we would get the same behavior. Why are you changing this file if it makes no difference?

Comment (WilburZjh) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:210]: Its not exactly the same, the Signature engines, Key Pair Generator engine, and Key Agreement engine are disabled when FIPS is enabled.

Comment (keithc-ca) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:210]: Ok, I see the differences now. The FIPS part is common to both branches: I suggest it be moved before an `if` statement:
```
        if (!FIPSConfigurator.enableFips()) {
            return;
        }
```
This will make it much clearer which services apply in FIPS mode or otherwise.

It would be good if you could make the distinction clearer in `SunEntries.java` as well.

Comment (WilburZjh) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java]: sure, I changed both `SunEntries.java` and `SunEC.java`

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 5]
```diff
@@ -109,6 +111,51 @@ public final class SunEntries {
         // common attribute map
         HashMap<String, String> attrs = new HashMap<>(3);
 
+        /*
+         * Certificates
+         */
+        addWithAlias(p, "CertificateFactory", "X.509",
+                "sun.security.provider.X509Factory", attrs);
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java:118]: The attributes are different than before and different than in jdk11. For example, for `CertificateFactory`, they include `ImplementedIn=Software`.

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java [hunk 6]
```diff
@@ -247,6 +255,10 @@ void putEntries() {
         putService(new ProviderServiceA(this, "AlgorithmParameters",
             "EC", "sun.security.util.ECParameters", apAttrs));
 
+        if (!FIPSConfigurator.enableFIPS()) {
```
Comment (keithc-ca) [inline src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:258]: Again, the method name is wrong and the condition is inverted.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/sun/security/provider/SunEntries.java (hunk 5 above)
Comment (WilburZjh) [reply src/java.base/share/classes/sun/security/provider/SunEntries.java:119]: added

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 7]
```diff
@@ -109,6 +111,51 @@ public final class SunEntries {
...
+        add(p, "CertPathBuilder", "PKIX",
+                "sun.security.provider.certpath.SunCertPathBuilder",
+                attrs);
+        add(p, "CertPathValidator", "PKIX",
+                "sun.security.provider.certpath.PKIXCertPathValidator",
+                attrs);
+
+        if (!FIPSConfigurator.enableFIPS()) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/java.base/share/classes/sun/security/provider/SunEntries.java]: modified

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java (hunk 6 above)
Comment (WilburZjh) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java]: modified

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java [hunk 8]
```diff
@@ -201,7 +209,7 @@ void putEntries() {
         HashMap<String, String> ATTRS = new HashMap<>(3);
         ATTRS.put("ImplementedIn", "Software");
         String ecKeyClasses = "java.security.interfaces.ECPublicKey" +
-                 "|java.security.interfaces.ECPrivateKey";
+                "|java.security.interfaces.ECPrivateKey";
```
Comment (keithc-ca) [inline src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java:212]: Please revert this change.

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 9]
```diff
@@ -246,12 +294,12 @@ public final class SunEntries {
             providerSHA256 = "sun.security.provider.SHA2$SHA256";
             providerSHA384 = "sun.security.provider.SHA5$SHA384";
             providerSHA512 = "sun.security.provider.SHA5$SHA512";
-        }
+		}
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java:297]: Spaces should be used for indentation.

File: src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java (hunk 8 above)
Comment (WilburZjh) [reply src/jdk.crypto.ec/share/classes/sun/security/ec/SunEC.java]: reverted

File: src/java.base/share/classes/sun/security/provider/SunEntries.java (hunk 9 above)
Comment (WilburZjh) [reply src/java.base/share/classes/sun/security/provider/SunEntries.java]: changed to spaces

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 10]
```diff
@@ -108,6 +110,54 @@ public final class SunEntries {
 
         // common attribute map
         HashMap<String, String> attrs = new HashMap<>(3);
+        if (FIPSConfigurator.enableFips()) {
+            // FIPS supported
+            attrs.put("ImplementedIn", "Software");
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java:115]: This should be included unconditionally.

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 11]
```diff
@@ -108,6 +110,54 @@ public final class SunEntries {
...
+         */
+        add(p, "Configuration", "JavaLoginConfig",
+                "sun.security.provider.ConfigFile$Spi");
+
+        /*
+         * CertPathBuilder and CertPathValidator
+         */
+        attrs.clear();
+        attrs.put("ValidationAlgorithm", "RFC5280");
+        attrs.put("ImplementedIn", "Software");
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java:149]: Lines 147 and 149 can be removed, `ImplementedIn -> Software` is the only entry in the map when we get here.

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 12]
```diff
@@ -108,6 +110,54 @@ public final class SunEntries {
...
+                attrs);
+        add(p, "CertPathValidator", "PKIX",
+                "sun.security.provider.certpath.PKIXCertPathValidator",
+                attrs);
+
+        if (FIPSConfigurator.enableFips()) {
+            return;
+        }
 
         /*
```
Comment (keithc-ca) [inline src/java.base/share/classes/sun/security/provider/SunEntries.java:162]: The `attr` map must be cleared here so the following have the same properties when not in FIPS mode.

File: closed/src/java.base/share/classes/openj9/internal/security/FIPSConfigurator.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        super();
+    }
+
+    /**
+     * FIPS mode will be enabled only if the semeru.fips system
+     * property is true (default as false).
+     *
+     * @return true if FIPS is enabled
+     */
+    public static boolean enableFips() {
```
Comment (keithc-ca) [inline closed/src/java.base/share/classes/openj9/internal/security/FIPSConfigurator.java:73]: Suggest renaming this method `enableFIPS`.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/sun/security/provider/SunEntries.java (hunk 10 above)
Comment (WilburZjh) [reply src/java.base/share/classes/sun/security/provider/SunEntries.java]: updated

File: src/java.base/share/classes/sun/security/provider/SunEntries.java (hunk 11 above)
Comment (WilburZjh) [reply src/java.base/share/classes/sun/security/provider/SunEntries.java]: removed

File: src/java.base/share/classes/sun/security/provider/SunEntries.java [hunk 14]
```diff
@@ -109,6 +111,53 @@ public final class SunEntries {
         // common attribute map
         HashMap<String, String> attrs = new HashMap<>(3);
 
+        // FIPS supported
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/java.base/share/classes/sun/security/provider/SunEntries.java]: updated

File: src/java.base/share/classes/java/security/SecureRandom.java [hunk 15]
```diff
@@ -267,23 +275,32 @@ public SecureRandom(byte[] seed) {
...
-            } else {
-                prngService = p.getDefaultSecureRandomService();
-                if (prngService != null) {
-                    prngAlgorithm = prngService.getAlgorithm();
+
+        // If in FIPS mode, use the SecureRandom from the FIPS provider
+        if (FIPSConfigurator.enableFIPS()) {
+            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
+            prngAlgorithm = "PKCS11";
+            prngService = p.getService("SecureRandom", prngAlgorithm);
```
[the remark this answers is not in the document]
Comment (WilburZjh) [reply src/java.base/share/classes/java/security/SecureRandom.java]: I update the code to check that `p != null` before calling `getService()`. For `prngService`, its a local variable and already be set as `null`, I dont think it need to be checked before setting `prngAlgoritm`.

File: src/java.base/share/classes/java/security/SecureRandom.java [hunk 16]
```diff
@@ -267,23 +275,34 @@ public SecureRandom(byte[] seed) {
...
         String prngAlgorithm = null;
-        for (Provider p : Providers.getProviderList().providers()) {
-            // SUN provider uses the SunEntries.DEF_SECURE_RANDOM_ALGO
-            // as the default SecureRandom algorithm; for other providers,
-            // Provider.getDefaultSecureRandom() will use the 1st
-            // registered SecureRandom algorithm
-            if (p.getName().equals("SUN")) {
-                prngAlgorithm = SunEntries.DEF_SECURE_RANDOM_ALGO;
+
+        // If in FIPS mode, use the SecureRandom from the FIPS provider
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/security/SecureRandom.java:279]: Please finish the sentence with a period.

File: src/java.base/share/classes/java/security/SecureRandom.java [hunk 17]
```diff
@@ -267,23 +275,34 @@ public SecureRandom(byte[] seed) {
...
+        if (FIPSConfigurator.enableFIPS()) {
+            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
+            prngAlgorithm = "PKCS11";
+            if (p != null) {
                 prngService = p.getService("SecureRandom", prngAlgorithm);
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/security/SecureRandom.java:284]: Should this not fail in some way if `prngService == null` if FIPS mode is requested?

File: src/java.base/share/classes/java/security/SecureRandom.java (hunk 16 above)
Comment (WilburZjh) [reply src/java.base/share/classes/java/security/SecureRandom.java]: added.

File: src/java.base/share/classes/java/security/SecureRandom.java (hunk 17 above)
Comment (WilburZjh) [reply src/java.base/share/classes/java/security/SecureRandom.java:284]: > 

I added two exceptions, one for not finding the `SunPKCS11-NSS-FIPS` provider, the other one for not getting the `PKCS11` correctly.

Comment (keithc-ca) [reply src/java.base/share/classes/java/security/SecureRandom.java:284]: Suggest inverting the if statements and adjusting the exception messages (which typically don't end with a period):
```
            if (p == null) {
                throw new RuntimeException("could not find SunPKCS11-NSS-FIPS provider for FIPS mode");
            }
            prngService = p.getService("SecureRandom", prngAlgorithm);
            if (prngService == null) {
                throw new RuntimeException("could not find SecureRandom implementation from SunPKCS11-NSS-FIPS");
            }
```

Comment (WilburZjh) [reply src/java.base/share/classes/java/security/SecureRandom.java]: inverted

File: src/java.base/share/classes/java/security/SecureRandom.java [hunk 18]
```diff
@@ -267,23 +275,38 @@ public SecureRandom(byte[] seed) {
...
+        if (FIPSConfigurator.enableFIPS()) {
+            Provider p = Security.getProvider("SunPKCS11-NSS-FIPS");
+            prngAlgorithm = "PKCS11";
+            if (p == null) {
+                throw new RuntimeException("could not find a implementation for PKCS11 in SunPKCS11-NSS-FIPS.");
+            }
+            prngService = p.getService("SecureRandom", prngAlgorithm);
+            if (prngService == null) {
+                throw new RuntimeException("could not find SunPKCS11-NSS-FIPS provider in FIPS mode.");
+            }
```
Comment (keithc-ca) [inline src/java.base/share/classes/java/security/SecureRandom.java:289]: Please don't include the period in the messages.

I think you have the messages swapped: the first should be about the missing provider, the second about the missing service.

Also correct the grammar (a -> an), or use my shorter suggestions:
```
    "could not find SunPKCS11-NSS-FIPS provider for FIPS mode"
```
and
```
   "could not find SecureRandom implementation from SunPKCS11-NSS-FIPS"
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/security/SecureRandom.java (hunk 18 above)
Comment (WilburZjh) [reply src/java.base/share/classes/java/security/SecureRandom.java]: updated

**Comment by keithc-ca** [comment]
jenkins test sanity,extended xlinux jdk17

**Review by keithc-ca — APPROVED** [review]
